pkg/ddl/ddl_running_jobs.go

```diff
@@ -37,57 +37,180 @@ type runningJobs struct {
+	schemas           map[string]map[string]struct{}
+	placementPolicies map[string]struct{}
+	resourceGroups    map[string]struct{}
+}
+
+func newObjects() *objects {
+	return &objects{
+		schemas:           make(map[string]map[string]struct{}),
+		placementPolicies: make(map[string]struct{}),
+		resourceGroups:    make(map[string]struct{}),
+	}
 }
 
-// TODO(lance6716): support jobs involving objects that are exclusive (ALTER
-// PLACEMENT POLICY vs ALTER PLACEMENT POLICY) and shared (ALTER TABLE PLACEMENT
-// POLICY vs ALTER TABLE PLACEMENT POLICY)
 func newRunningJobs() *runningJobs {
 	return &runningJobs{
 		ids:          make(map[int64]struct{}),
 		idsStrGetter: func() string { return "" },
-		schemas:      make(map[string]map[string]struct{}),
+		exclusive:    newObjects(),
+		shared:       newObjects(),
+		pending:      newObjects(),
 	}
 }
 
-// add should only add the argument that passed the last checkRunnable.
-func (j *runningJobs) add(jobID int64, involves []model.InvolvingSchemaInfo) {
+// checkRunnable checks whether the job can be run. If the caller found a
+// runnable job and decides to add it, it must addRunning before next
+// checkRunnable invocation. Otherwise, it should addPending before next
+// checkRunnable invocation.
+func (j *runningJobs) checkRunnable(jobID int64, involves []model.InvolvingSchemaInfo) bool {
+	j.mu.RLock()
+	defer j.mu.RUnlock()
+
+	if _, ok := j.ids[jobID]; ok {
+		// should not happen
+		if intest.InTest {
+			panic(fmt.Sprintf("job %d is already running", jobID))
+		}
+		return false
+	}
+	// Currently flashback cluster is the only DDL that involves ALL schemas.
+	if _, ok := j.exclusive.schemas[model.InvolvingAll]; ok {
+		return false
+	}
+
+	for _, info := range involves {
+		if intest.InTest {
+			if info.Database == model.InvolvingNone && info.Table != model.InvolvingNone {
+				panic(fmt.Sprintf(
+					"job %d is invalid. While database is empty, involved table name is not empty: %s",
+					jobID, info.Table,
+				))
+			}
+			if info.Database != model.InvolvingNone && info.Table == model.InvolvingNone {
+				panic(fmt.Sprintf(
+					"job %d is invalid. While table is empty, involved database name is not empty: %s",
+					jobID, info.Database,
+				))
+			}
+		}
+
+		if info.Database == model.InvolvingAll && info.Table == model.InvolvingAll &&
+			info.Mode == model.ExclusiveInvolving {
+			// check for involving all databases and tables, where the only case is FLASHBACK
+			// CLUSTER
+			if len(j.exclusive.schemas) != 0 || len(j.shared.schemas) != 0 ||
+				len(j.exclusive.placementPolicies) != 0 || len(j.shared.placementPolicies) != 0 ||
+				len(j.exclusive.resourceGroups) != 0 || len(j.shared.resourceGroups) != 0 {
+				return false
+			}
+			continue
+		}
+
+		var toCheck []*objects
+		switch info.Mode {
+		case model.ExclusiveInvolving:
+			// Exclusive objects conflicts with running exclusive and shared objects.
+			toCheck = []*objects{j.exclusive, j.shared}
```

Comment:
check `pending`?

Reply:
I think `pending` is caused by previous `shared` objects, so check `shared` is enough. But then I reallised `shared` objects may be removed concurrently during a round of `getJob`. I'll add `pending` in next commit